Llk refactor uplift gs

rtawfik01 opened the pull request.

muthutt reviewed (commented):
I'm trying to run some tests on GH and WH manually and inspect your code; should have review by end of day and you should be open to submit on Tuesday this week (tomorrow)

muthutt reviewed (approved):
LGTM - to unblock large PR - but request to address minor issues

rtawfik01 commented on tt_metal/hw/ckernels/wormhole_b0/common/inc/ckernel.h:
```diff
@@ -36,7 +36,7 @@
 #endif
 
-#ifdef LLK_TB_TEST
+#if defined(EN_KERNEL_SLOWDOWN)
```
Hi @muthutt , its because the rest of the defines all use `#ifndef`, so changing the `#ifdef` to `#if defined` provides a more visible difference than using `ifdef` and `ifndef` sequentially in the same file (just easier for debugging). If its an issue though for the coding standard I can definitely change it.

rtawfik01 replied:
Hey @muthutt , since the above file is common between buda & metal and will be in its own repo soon, we will keep that define to keep it parity until the common submodule

rtawfik01 committed:
#4270: Uplift GS llk_lib & ckernel files/ Add llk_api files

acejkov committed:
#4270: Replaced enum dim with vector_mode in llk core library

acejkov committed:
#4270: Moved ckernel_ops.h to common/inc and removed unused header llk_format_conversions.h from llk core lib

rtawfik01 committed:
#4270: Fixes for GS compile/regressions
	- Remove llk_format_conversions includes
	- Fix compute_kernel_api files to call new apis
	- Add missing pack_untilize functions
	- Rename reg_read_barrier to reg_read
	- Add missing llk_setup_operands impl

acejkov committed:
#4270: Replaced var mul with MUL_TILE_SIZE_AND_INDEX macro. removed save/restore for unpack untilize init/unint

acejkov committed:
#4270: Fixed tile offset addressing in unpack AB matmul

mywoodstock commented on tt_metal/hw/ckernels/grayskull/common/inc/ckernel_globals.h:
```diff
@@ -12,7 +12,11 @@ extern uint32_t cfg_state_id;
 extern uint32_t gl_alu_format_spec_reg;
 
-extern volatile uint32_t l1_buffer[16];
+extern volatile uint32_t tt_l1_ptr l1_buffer[16];
```
should this be a pointer?

rtawfik01 replied:
This l1 buffer is used by the reduce op and I believe it meant to be static l1 memory, but Ill add @acejkov to confirm.

acejkov replied:
yes, we need it for buda

DrJessop commented on tt_metal/hw/ckernels/grayskull/llk_lib/llk_math_eltwise_binary.h:
```diff
@@ -3,75 +3,81 @@
+                TT_ZEROACC(p_zeroacc::CLR_16, ADDR_MOD_1, (math_sync_tile_dst_index << 2) + 2);
+                TT_ZEROACC(p_zeroacc::CLR_16, ADDR_MOD_1, (math_sync_tile_dst_index << 2) + 3);
             }
+            */
```
please remove comments.

rtawfik01 replied:
Done

DrJessop commented on tt_metal/hw/ckernels/grayskull/llk_lib/llk_pack_common.h:
```diff
@@ -1,42 +1,52 @@
+    }
+#else
     TTI_SEMWAIT(p_stall::STALL_TDMA, semaphore::t6_sem(semaphore::MATH_PACK), p_stall::STALL_ON_ZERO);
+#endif
```
Please remove all perf dump code.

rtawfik01 replied:
answered in the other perf dump comment

DrJessop commented on tt_metal/hw/ckernels/grayskull/llk_lib/llk_pack_common.h:
```diff
@@ -138,6 +184,74 @@ inline void llk_pack_relu_config(std::uint32_t config) {
+            TTI_WRCFG(p_gpr_pack::TMP0,  p_cfg::WRCFG_32b, TILE_ROW_SET_MAPPING_1_row_set_mapping_0_ADDR32);
+            TTI_WRCFG(p_gpr_pack::TMP0,  p_cfg::WRCFG_32b, TILE_ROW_SET_MAPPING_2_row_set_mapping_0_ADDR32);
+            TTI_WRCFG(p_gpr_pack::TMP0,  p_cfg::WRCFG_32b, TILE_ROW_SET_MAPPING_3_row_set_mapping_0_ADDR32);
+        } else {
```
Not following why this is templatized on untilize. If it's just for the sake of bringing this up to date with latest buda backend, that's fine I guess.

rtawfik01 replied:
yes this is common with buda, but if its not used we will clean it up there too

DrJessop commented on tt_metal/hw/ckernels/grayskull/llk_lib/llk_unpack_AB.h:
```diff
@@ -170,4 +135,8 @@ inline void llk_unpack_AB(
+
+#ifdef PERF_DUMP
+    first_unpack_recorded = true;
+#endif
```
Will mention one more time, but this should apply everywhere. There should not be any perf dump code blocks, we don't support the flow used in budabackend.

rtawfik01 replied:
agree we will have to remove perf dump code, but problem is right now is that buda perf dump hooks are ingrained in the llk_lib kernels, and will cause all buda perf test sweeps to fail if removed now (because this file will be parity with buda). But I agree @acejkov and I will review and separate these perf hooks to be in the buda llk_api files once these files are commonalized and we can make sure buda regressions don't fail.

acejkov replied:
there might be an option to reuse perf dump tools for metal. we keep as integral part of the LLK and higher level stack can decide to use it or not

mywoodstock commented on tt_metal/hw/ckernels/grayskull/llk_lib/llk_pack_untilize.h:
```diff
@@ -70,29 +68,3 @@ inline void _llk_pack_untilize_(const std::uint32_t address, const std::uint32_t
-
-        pack_tile_addr += block_ct_dim*(std::uint32_t)(GET_L1_HEADERLESS_TILE_SIZE(pack_dst_format[output_id]));
-    }
-}
```
these have been removed?

mywoodstock replied:
ok nvm, i see its moved :)

rtawfik01 replied:
yes they are in llk_pack_api.h now

DrJessop reviewed (approved):
Approving under the following condition:
- Remove any instances of perf dump code
- Remove commented out code

I didn't spend time reviewing code quality since this just seems like a straight up uplift to have up to date budabackend code, after which we'll be commonizing the two stacks. Let me know if you want a more thorough review if that wasn't the only scope of this PR.

mywoodstock reviewed (approved):
lgtm!

rtawfik01 commented on tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_negative.h:
```diff
@@ -0,0 +1,31 @@
+
+template <bool APPROXIMATE, DstSync Dst = DstSync::SyncFull>
+inline void llk_math_eltwise_unary_sfpu_negative(uint dst_index, int vector_mode = (int)VectorMode::RC) {
+    constexpr int first_iterations = 1;
```
Fixed, first iteration was removed in the other PR for wormhole_b0

acejkov committed:
#4270: added missing edge mask setting for reduce_scaler

rtawfik01 committed:
#4270: Clean up comments/remove unused includes

acejkov committed:
#4270: Clear edge mask 1 in _llk_pack_reduce_mask_clear_ fun

rtawfik01 committed:
#4270: Revert to sfpu sign function before uplift

acejkov committed:
#4270: Removed unused function mem_barrier

rtawfik01 committed:
#4270: Fix compile error for untilize binary

davorchap commented on tt_metal/hw/ckernels/grayskull/common/inc/ckernel.h:
```diff
@@ -267,36 +332,169 @@ inline void cfg_rmw_gpr(uint32_t cfg_addr32, uint32_t cfg_shamt, uint32_t cfg_ma
     cfg_rmw(cfg_addr32, cfg_shamt, cfg_mask, wrdata);
 }
 
+inline void mailbox_write(const uint8_t thread, const uint32_t data)
```
Are these use in the new set of GS LLKs? Or are these for debug infra?

rtawfik01 replied:
These are for debug infra

davorchap commented on tt_metal/hw/ckernels/grayskull/common/inc/ckernel.h:
```diff
@@ -267,36 +332,169 @@ inline void cfg_rmw_gpr(uint32_t cfg_addr32, uint32_t cfg_shamt, uint32_t cfg_ma
+// If PERF_DUMP is enabled, always wait but only for the inputs that perf dump is enabled for
+// If PERF_DUMP is enabled, and delay is not, no need to insert these apis for unpack and math
+template<int thread_id>
+inline void serialize_input_loop_start() {
```
The PERF_DUMP infra -- this is BBE specific?
We can have it for now in `ckernel.h` , but perhaps it can be factored out to its own file later on, so that if not used we don't have to compile it.

rtawfik01 replied:
PERF_DUMP is currently used by BBE, but the perf infra in the llks can be re-used by either stacks if required since it is general

davorchap commented on tt_metal/hw/ckernels/grayskull/common/inc/ckernel_addr_map.h:
```diff
@@ -0,0 +1,14 @@
+// SPDX-License-Identifier: Apache-2.0
+
+// Temp address map in bytes
+#define TRISC0_PARAMS_BASE_ADDRESS 32*1024
```
If also BBE specific, we should plan how to factor out to a separate file.

davorchap commented on tt_metal/hw/ckernels/grayskull/llk_lib/llk_unpack_AB_matmul.h:
```diff
@@ -70,41 +62,29 @@ ALWI void llk_unpack_AB_matmul_mop_config(bool transpose) {
-    std::uint32_t address_a = base_address_a + offset_address_a - 1;
-    std::uint32_t address_b = base_address_b + offset_address_b - 1;
+    // Todo: do something with tile dim flags
+    volatile uint tt_reg_ptr *cfg = get_cfg_pointer();  // get pointer to registers for current state ID
```
I believe hoisting this outside the function and before the matmul rt_dim/ct_dim loops did improve performance. 
@mywoodstock measured BERT-L on this branch at 411 sent/sec, so perhaps it's getting auto-hoisted in the refactored code.

rtawfik01 replied:
I think when we commonlize the code, we can try taking the cfg pointer specifically out and see if it helps performance for metal/buda, but I think the bigger impact was the address calculations being in the higher level loops:
```
    const std::uint32_t operandA_id = get_operand_id(operandA);
    const std::uint32_t operandB_id = get_operand_id(operandB);

    std::uint32_t base_address_a = cb_interface[operandA_id].fifo_rd_ptr - 1;
    std::uint32_t base_address_b = cb_interface[operandB_id].fifo_rd_ptr - 1;
    std::uint32_t unpA_src_format = (std::uint32_t)unpack_src_format[operandA_id];
    std::uint32_t unpB_src_format = (std::uint32_t)unpack_src_format[operandB_id];

    for (std::uint32_t rt=0; rt<rt_dim; rt++) {
        std::uint32_t offset_address_a = MUL_TILE_SIZE_AND_INDEX<true>(unpA_src_format, (tile_index_a + rt*kt_dim));
        std::uint32_t address_a = base_address_a + offset_address_a;

        for (std::uint32_t ct=0; ct<ct_dim; ct++) {

            std::uint32_t offset_address_b = MUL_TILE_SIZE_AND_INDEX<true>(unpB_src_format, (tile_index_b+ct));
            std::uint32_t address_b = base_address_b + offset_address_b;

            _llk_unpack_AB_matmul_(
                address_a,
                address_b
            );
```

because we did see before that being able to push those address calculations to replay buffers/outerloops in wormhole gave a lot of performance gains. 

But currently because llk_unpack_AB_matmul.h file is common for both buda/metal, and buda is still calculating by tile, then the cfg pointer will have to stay in the per tile calculation until we commonlize/back-propagate back to buda

davorchap commented on tt_metal/hw/ckernels/grayskull/llk_lib/llk_unpack_AB_matmul.h:
```diff
@@ -139,140 +107,8 @@ inline void llk_unpack_AB_matmul(
-    std::uint32_t base_address_a = cb_interface[inputA].fifo_rd_ptr - 1;
-    std::uint32_t base_address_b = cb_interface[inputB].fifo_rd_ptr - 1;
-
-    volatile uint tt_reg_ptr *cfg = get_cfg_pointer();  // get pointer to registers for current state ID
```
hoisting of the `cfg` ptr outside of the loops in the old implementation

rtawfik01 commented on tt_metal/hw/ckernels/grayskull/common/inc/ckernel_pcbuf.h:
```diff
@@ -0,0 +1,46 @@
+namespace ckernel
+{
+
+inline uint get_pc_buf_cmd(uint ckernel_id, uint command_addr)
```
Yes it is in the firmware

davorchap commented:
overall it looks great, thank you!

* made comments about some code that is not used in Metal, so perhaps could be refactor into separate files so that compile can be skipped
* llk_matmul_tile in the block mode (rt_dim, ct_dim) has changed, ie `cfg` ptr is not manually hoisted out of the rt/ct loops, but there doesn't seem to be a perf drop based on Abhinav's measurements. I'm not sure `cfg` can  be auto-hoisted given that it is volatile.


@mywoodstock measurements: 
ResNet50-b20 
"AVG DEVICE KERNEL SAMPLES/S": 6156.77346282875,
    "MIN DEVICE KERNEL SAMPLES/S": 6152.297667848523,
    "MAX DEVICE KERNEL SAMPLES/S": 6164.067125458182,
 
BERT-L, b12
 "AVG DEVICE KERNEL SAMPLES/S": 411.0532633569959,
    "MIN DEVICE KERNEL SAMPLES/S": 410.89907045042617,
    "MAX DEVICE KERNEL SAMPLES/S": 411.16637415605527,
